pennylane/measurements/__init__.py

```diff
@@ -11,8 +11,121 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""
-This module contains measurements supported by PennyLane.
+r"""
+This module contains all the measurements supported by PennyLane.
+
+Description
+-----------
+
+Measurements
+~~~~~~~~~~~~
+The :class:`MeasurementProcess` class serves as a base class for measurements, and is inherited
+from the :class:`SampleMeasurement`, :class:`StateMeasurement` and :class:`MeasurementTransform`
+classes. These classes are subclassed to implement measurements in PennyLane.
+
+* Each :class:`SampleMeasurement` subclass represents a sample-based measurement, which contains a
+  :func:`SampleMeasurement.process_samples` method that processes the sequence of samples generated
+  by the device. See :class:`CountsMP` for an example.
+
+* Each :class:`StateMeasurement` subclass represents a state-based measurement, which contains a
+  :func:`StateMeasurement.process_state` method that processes the quantum state generated by the
+  device. See :class:`StateMP` for an example.
+
+* Each :class:`MeasurementTransform` subclass represents a measurement process that requires
+  the application of a batch transform, which converts the given quantum script into a batch of
+  quantum scripts and executes them using the device. The main difference between a
+  :class:`MeasurementTransform` and a :func:`~pennylane.batch_transform` is that a batch transform is
+  tracked by the gradient transform, while a :class:`MeasurementTransform` process isn't.
+  See :class:`ClassicalShadowMP` for an example.
+
+.. note::
+    A measurement process can inherit from both :class:`SampleMeasurement` and
+    :class:`StateMeasurement` classes, defining the needed logic to process either samples or the
+    quantum state. See :class:`VarianceMP` for an example.
+
+Differentiation
+^^^^^^^^^^^^^^^
+In general, a :class:`MeasurementProcess` is differentiable with respect to a parameter if the domain of
+that parameter is continuous. When using the analytic method of differentiation the output of the
+measurement process must be a real scalar value for it to be differentiable.
+
+Creating custom measurements
+----------------------------
+A custom measurement process can be created by inheriting from any of the classes mentioned above.
+
+The following is an example for a sample-based measurement that computes the number of samples
+obtained of a given state:
+
+.. code-block:: python
+
+    import pennylane as qml
+    from pennylane.measurements import SampleMeasurement
+
+    class CountState(SampleMeasurement):
+        def __init__(self, state: str):
+            self.state = state  # string identifying the state e.g. "0101"
+            wires = list(range(len(state)))
+            super().__init__(wires=wires)
+
+        def process_samples(self, samples, wire_order, shot_range, bin_size):
```

Comment:
I reckon this signature cannot be arbitrary? If that is the case, would be important to stress that somewhere in the text!

Reply:
Done!